modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryEventBuffer.java

```diff
@@ -26,40 +26,69 @@
 import java.util.TreeMap;
 import java.util.concurrent.ConcurrentLinkedDeque;
 import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
+import org.apache.ignite.IgniteLogger;
 import org.apache.ignite.IgniteSystemProperties;
 import org.apache.ignite.internal.processors.affinity.AffinityTopologyVersion;
+import org.apache.ignite.internal.util.GridAtomicLong;
+import org.apache.ignite.internal.util.typedef.internal.LT;
 import org.apache.ignite.util.deque.FastSizeDeque;
 import org.jetbrains.annotations.Nullable;
 
 /**
  *
  */
 public class CacheContinuousQueryEventBuffer {
-    /** */
+    /** Maximum size of buffer for pending events. Default value is {@code 10_000}. */
+    public static final int MAX_PENDING_BUFF_SIZE =
+        IgniteSystemProperties.getInteger("IGNITE_CONTINUOUS_QUERY_PENDING_BUFF_SIZE", 10_000);
+
+    /** Batch buffer size. */
     private static final int BUF_SIZE =
         IgniteSystemProperties.getInteger("IGNITE_CONTINUOUS_QUERY_SERVER_BUFFER_SIZE", 1000);
 
     /** */
     private static final Object RETRY = new Object();
 
-    /** */
-    protected final int part;
+    /** Continuous query category logger. */
+    private final IgniteLogger log;
 
     /** */
-    private AtomicReference<Batch> curBatch = new AtomicReference<>();
+    private final int part;
 
-    /** */
-    private FastSizeDeque<CacheContinuousQueryEntry> backupQ = new FastSizeDeque<>(new ConcurrentLinkedDeque<>());
+    /** Batch of entries currently being collected to send to the remote. */
+    private final AtomicReference<Batch> curBatch = new AtomicReference<>();
 
-    /** */
-    private ConcurrentSkipListMap<Long, CacheContinuousQueryEntry> pending = new ConcurrentSkipListMap<>();
+    /** Queue for keeping backup entries which partition counter less the counter processing by current batch. */
+    private final FastSizeDeque<CacheContinuousQueryEntry> backupQ = new FastSizeDeque<>(new ConcurrentLinkedDeque<>());
+
+    /** Entries which are waiting for being processed. */
+    private final ConcurrentSkipListMap<Long, CacheContinuousQueryEntry> pending = new ConcurrentSkipListMap<>();
+
+    /**
+     * The size method of the pending ConcurrentSkipListMap is not a constant-time operation. Since each
```

Comment:
`size()` method is almost constant time operation, there is a long adder inside and usually, there is only one cell in it (but at most CPU count cells). If you have low contention on `pending` map access to `pending.size()` will cost relatively low. If you have very high contention on `pending` map, I think `AtomicInteger.incrementAndGet()` can become a bottleneck here. 
So I'm not sure about this optimization, perhaps it's better to use `pending.size()`.

Reply:
Discussed privately. jdk8 and jdk11 has different `size()` method implementation, so for the main use-case when the buffer is overflowed it's better to use AtomicInteger + synchronization on `pending` to eliminate all potential issues.